### PR TITLE
Propagate context and ensure git commands run in request context

### PR DESCRIPTION
**Pull request by zeripath**
This PR continues the work in #17125 by progressively ensuring that git
commands run within the request context.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: services/pull/pull.go [hunk 1]
```diff
@@ -52,7 +52,7 @@ func NewPullRequest(repo *models.Repository, pull *models.Issue, labelIDs []int6
 	pull.PullRequest = pr
 
 	if pr.Flow == models.PullRequestFlowGithub {
-		err = PushToBaseRepo(pr)
+		err = PushToBaseRepo(ctx, pr)
```
Comment (zeripath) [inline services/pull/pull.go]: This may not be correct here - I will have to think...

File: services/repository/branch.go [hunk 2]
```diff
@@ -60,32 +61,34 @@ func GetBranch(repo *models.Repository, branch string) (*git.Branch, error) {
...
+func GetBranches(ctx context.Context, repo *models.Repository, skip, limit int) ([]*git.Branch, int, error) {
+	return git.GetBranchesByPath(ctx, repo.RepoPath(), skip, limit)
 }
 
 // checkBranchName validates branch name with existing repository branches
-func checkBranchName(repo *models.Repository, name string) error {
-	gitRepo, err := git.OpenRepository(repo.RepoPath())
+// FIXME: There really has to be faster mechanism than this. git cat-file --batch-check will check a branchname
+// and or tag name... (git.DefaultContext - as a marker to come back to this)
+func checkBranchName(ctx context.Context, repo *models.Repository, name string) error {
```
Comment (zeripath) [inline services/repository/branch.go]: This function is very inefficient. I've managed to improve it already but it could be a lot quicker.

**Comment by zeripath** [comment]
OK all of the calls to git now propagate the context and avoid reopening the gitrepo.

Need to ensure that transactional correctness is ensured especially with regards to post-receive.

File: models/review.go [hunk 3]
```diff
@@ -155,8 +155,8 @@ func (r *Review) loadAttributes(ctx context.Context) (err error) {
 }
 
 // LoadAttributes loads all attributes except CodeComments
-func (r *Review) LoadAttributes() error {
-	return r.loadAttributes(db.DefaultContext)
+func (r *Review) LoadAttributes(ctx context.Context) error {
+	return r.loadAttributes(ctx)
```
Comment (lunny) [inline models/review.go]: So we can remove `loadAttributes`

File: models/review.go [hunk 4]
```diff
@@ -98,8 +98,8 @@ func (r *Review) loadCodeComments(ctx context.Context) (err error) {
 }
 
 // LoadCodeComments loads CodeComments
-func (r *Review) LoadCodeComments() error {
-	return r.loadCodeComments(db.DefaultContext)
+func (r *Review) LoadCodeComments(ctx context.Context) error {
+	return r.loadCodeComments(ctx)
```
Comment (KN4CK3R) [inline models/review.go]: `loadCodeComments` is not needed anymore.

File: modules/git/blame.go [hunk 5]
```diff
@@ -114,11 +114,15 @@ func (r *BlameReader) Close() error {
...
-	if err != nil {
-		return nil, err
+	gitRepo := RepositoryFromContext(ctx, repoPath)
+	if gitRepo == nil {
+		var err error
+		gitRepo, err = OpenRepositoryCtx(ctx, repoPath)
+		if err != nil {
+			return nil, err
+		}
+		gitRepo.Close()
```
Comment (KN4CK3R) [inline modules/git/blame.go]: Is this a "repo exists" check?

**Comment by KN4CK3R** [comment]
There are multiple similar code blocks like:
```go
gitRepo := git.RepositoryFromContext(ctx, repoPath)
if gitRepo == nil {
	gitRepo, err = git.OpenRepositoryCtx(ctx, repoPath)
	if err != nil {
		return err
	}
	defer gitRepo.Close()
}
```
It may be worth to move that into a helper method like:
```go
func GetOrOpenRepository(ctx context.Context, path string) (*git.Repository, io.Closer, error) {
	gitRepo := git.RepositoryFromContext(ctx, path)
	if gitRepo != nil {
		return gitRepo, nil, nil
	}
	
	gitRepo, err := git.OpenRepositoryCtx(ctx, path)
	if err != nil {
		return nil, nil, err
	}
	return gitRepo, gitRepo, err
}

repo, closable, err := git.GetOrOpenRepository(ctx, "...")
if err != nil {
	...
}
if closable != nil {
	defer closable.Close()
}
```

File: modules/git/blame.go (hunk 5 above)
Comment (zeripath) [reply modules/git/blame.go]: Yeah, but it's not a new check.

Comment (KN4CK3R) [reply modules/git/blame.go]: Yes, it's not new but is it necessary?

**Comment by zeripath** [comment]
We could actually just return an empty Closer instead then we always run.

I did consider returning a deferrable (eg. context.CancelFunc) but thought it was too confusing.

File: modules/git/blame.go (hunk 5 above)
Comment (zeripath) [reply modules/git/blame.go]: It appears unnecessary. I've removed.

File: models/issue_comment.go [hunk 6]
```diff
@@ -705,11 +705,11 @@ func (c *Comment) LoadPushCommits() (err error) {
 		c.NewCommit = data.CommitIDs[1]
 	} else {
 		repoPath := c.Issue.Repo.RepoPath()
-		gitRepo, err := git.OpenRepository(repoPath)
+		gitRepo, closer, err := git.RepositoryFromContextOrOpen(ctx, repoPath)
```
Comment (6543) [inline models/issue_comment.go:749]: just wonder what's the difference between `OpenRepositoryCtx` and `RepositoryFromContextOrOpen` ?

Comment (6543) [reply models/issue_comment.go:749]: never mid :swan:

File: modules/context/api.go [hunk 7]
```diff
@@ -303,6 +303,7 @@ func APIContexter() func(http.Handler) http.Handler {
 			ctx.Resp.Header().Set(`X-Frame-Options`, setting.CORSConfig.XFrameOptions)
 
 			ctx.Data["CsrfToken"] = html.EscapeString(ctx.csrf.GetToken())
+			ctx.Data["Context"] = &ctx
```
Comment (6543) [inline modules/context/api.go:306]: do we realy need the context for api requests injected? (api do not have templates - but I guess we still use it in frontend somewhere ? :/ )

File: modules/context/api.go [hunk 8]
```diff
@@ -321,35 +322,32 @@ func APIContexter() func(http.Handler) http.Handler {
...
+	return func(ctx *APIContext) (cancel context.CancelFunc) {
+		// Empty repository does not have reference information.
+		if !allowEmpty && ctx.Repo.Repository.IsEmpty {
+			return
+		}
+
+		// For API calls.
+		if ctx.Repo.GitRepo == nil {
+			repoPath := repo_model.RepoPath(ctx.Repo.Owner.Name, ctx.Repo.Repository.Name)
+			gitRepo, err := git.OpenRepositoryCtx(ctx, repoPath)
```
Comment (6543) [inline modules/context/api.go:335]: I think we can use RepositoryFromContextOrOpen here

File: modules/git/repo_base.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	if repo, ok := value.(*Repository); ok && repo != nil {
+		if repo.Path == path {
+			return repo
+		}
+	}
+
+	return nil
+}
+
+type nopCloser func()
```
Comment (6543) [inline modules/git/repo_base.go:36]: I'm aware of type X struct{} - does a func() type have any specific means?

File: modules/notification/action/action.go [hunk 10]
```diff
@@ -389,7 +394,7 @@ func (a *actionNotifier) NotifySyncPushCommits(pusher *user_model.User, repo *re
 	}
 }
 
-func (a *actionNotifier) NotifySyncCreateRef(doer *user_model.User, repo *repo_model.Repository, refType, refFullName string) {
+func (a *actionNotifier) NotifySyncCreateRef(doer *user_model.User, repo *repo_model.Repository, refType, refFullName, shaSum string) {
```
Comment (6543) [inline modules/notification/action/action.go]:
```suggestion
func (a *actionNotifier) NotifySyncCreateRef(doer *user_model.User, repo *repo_model.Repository, refType, refFullName, refID string) {
```

File: modules/notification/action/action.go [hunk 11]
```diff
@@ -330,7 +335,7 @@ func (a *actionNotifier) NotifyPushCommits(pusher *user_model.User, repo *repo_m
 	}
 }
 
-func (a *actionNotifier) NotifyCreateRef(doer *user_model.User, repo *repo_model.Repository, refType, refFullName string) {
+func (a *actionNotifier) NotifyCreateRef(doer *user_model.User, repo *repo_model.Repository, refType, refFullName, shaSum string) {
```
Comment (6543) [inline modules/notification/action/action.go]:
```suggestion
func (a *actionNotifier) NotifyCreateRef(doer *user_model.User, repo *repo_model.Repository, refType, refFullName, refID string) {
```

File: modules/context/api.go (hunk 7 above)
Comment (zeripath) [reply modules/context/api.go:306]: Markdown render

File: modules/context/api.go (hunk 8 above)
Comment (zeripath) [reply modules/context/api.go:335]: The close function here is different

File: modules/git/repo_base.go (hunk 9 above)
Comment (zeripath) [reply modules/git/repo_base.go:36]: it's essentially the same as struct{} here except that it should more easily get inlined to a noop.

However, say instead you wanted to wrap some function to become an io.Closer, you could write:

```go
type asCloser func() error

func (a asCloser) Close() error {
  return a()
} 
```

and avoid the wrapper struct.

File: models/issue_comment.go [hunk 12]
```diff
@@ -724,7 +724,7 @@ func (c *Comment) CodeCommentURL() string {
 }
 
 // LoadPushCommits Load push commits
-func (c *Comment) LoadPushCommits() (err error) {
+func (c *Comment) LoadPushCommits(ctx context.Context) (err error) {
```
Comment (lunny) [inline models/issue_comment.go:727]: This method should be moved to `service` package.

Comment (zeripath) [reply models/issue_comment.go:727]: The purpose of this very large PR is to refactor only to begin passing context down the callstack - I think that's enough changes without adding in other major refactors.

File: routers/private/default_branch.go [hunk 13]
```diff
@@ -34,38 +33,19 @@ func SetDefaultBranch(ctx *gitea_context.PrivateContext) {
...
 		if !git.IsErrUnsupportedVersion(err) {
-			gitRepo.Close()
+			ctx.Repo.GitRepo.Close()
 			ctx.JSON(http.StatusInternalServerError, private.Response{
 				Err: fmt.Sprintf("Unable to set default branch on repository: %s/%s Error: %v", ownerName, repoName, err),
 			})
 			return
 		}
 	}
-	gitRepo.Close()
```
Comment (6543) [inline routers/private/default_branch.go:66]: missing ctx.Repo.GitRepo.Close()

or will it be closed by some middle ware?

Comment (zeripath) [reply routers/private/default_branch.go:66]: it'll be closed by the middleware yes

File: routers/private/default_branch.go [hunk 14]
```diff
@@ -34,38 +33,19 @@ func SetDefaultBranch(ctx *gitea_context.PrivateContext) {
...
-			Err: fmt.Sprintf("Failed to get git repository: %s/%s Error: %v", ownerName, repoName, err),
-		})
-		return
-	}
-	if err := gitRepo.SetDefaultBranch(repo.DefaultBranch); err != nil {
+	ctx.Repo.Repository.DefaultBranch = branch
+	if err := ctx.Repo.GitRepo.SetDefaultBranch(ctx.Repo.Repository.DefaultBranch); err != nil {
 		if !git.IsErrUnsupportedVersion(err) {
-			gitRepo.Close()
+			ctx.Repo.GitRepo.Close()
```
Comment (6543) [inline routers/private/default_branch.go]:
```suggestion
```

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
I vote for milestone v1.16 !!!

**Review by Gusted — APPROVED** [review]
LGTM in general, wasn't able to spot a incorrect passed ctx(which is the only concern that could rise here).

**Comment by lafriks** [comment]
Naming should probably made more consistent: `OpenRepositoryCtx`, `NewCommandContext`, `CloneWithContext` (I mean Ctx, Context, WIthContext)

**Comment by zeripath** [comment]
Well hopefully we'll be able to drop all of the non-contexted variants very soon.

**Pull request merged**